File: lander/spring.py
# uncomment the next line if running in a notebook
# %matplotlib inline
import numpy as np
import matplotlib.pyplot as plt

g_m = 1
g_k = 1
g_x = 0
g_v = 1

# simulation time, timestep and time
g_t_max = 100
g_dt = 0.001
g_t_array = np.arange(0, g_t_max, g_dt)


def Euler():
    
    # mass, spring constant, initial position and velocity
    m = g_m
    k = g_k
    x = g_x
    v = g_v

    # simulation time, timestep and time
    t_max = g_t_max
    dt = g_dt
    t_array = np.arange(0, g_t_max, g_dt)

    # initialise empty lists to record trajectories
    x_list = []
    v_list = []

    # Euler integration
    for t in t_array:

        # append current state to trajectories
        x_list.append(x)
        v_list.append(v)

        # calculate new position and velocity
        a = -k * x / m
        x = x + dt * v
        v = v + dt * a
    # convert trajectory lists into arrays, so they can be sliced (useful for Assignment 2)
    x_array = np.array(x_list)
    v_array = np.array(v_list)
    return x_array, v_array

def Verlet():
    
    # mass, spring constant, initial position and velocity
    m = g_m
    k = g_k
    x = g_x
    v = g_v

    # simulation time, timestep and time
    t_max = g_t_max
    dt = g_dt
    t_array = np.arange(0, t_max, dt)


    xplus1 = g_x + v * dt

    # initialise empty lists to record trajectories
    x_list = []
    v_list = []
    
    for t in range(1,len(t_array)):
        if not x_list:
            x_list.append(x)
            x_list.append(xplus1)
            # v_list.append(v)
            continue
        x_list.append(x)
        
        a = -k * x / m
        x = 2 * x - x_list[t-1] + dt**2 * a
    
    return x_list
        

x_array, v_array = Euler()
vx_array = Verlet()

# plot the position-time graph
plt.figure(1)
plt.clf()
plt.xlabel('time (s)')
plt.grid()
plt.plot(g_t_array, x_array, label='x (m)')
plt.plot(g_t_array, v_array, label='v (m/s)')
plt.plot(g_t_array, vx_array, label='verlet x (m/s)')
plt.legend()
plt.show()

File: lander/spring.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <chrono>

using namespace std;

int main() {

  auto t1 = std::chrono::high_resolution_clock::now();

  // declare variables
  double m, k, x, v, xplus1, vplus1, t_max, dt, t, a;
  vector<double> t_list, x_list, v_list;

  // mass, spring constant, initial position and velocity
  m = 1;
  k = 1;
  x = 0;
  v = 1;


  // simulation time and timestep
  t_max = 1000000;
  dt = 0.1;

  xplus1 = x + v * dt;
  vplus1 = v;

  // Euler integration
  for (t = 0; t <= t_max; t = t + dt) {

    if (x_list.empty()) {
        x_list.push_back(x);
        v_list.push_back(v);
        x_list.push_back(xplus1);
        v_list.push_back(vplus1);
    }
    else {
        // append current state to trajectories
        t_list.push_back(t);
        x_list.push_back(x);
        v_list.push_back(v);

        // calculate new position and velocity
        a = -k * x / m;
        x = 2 * x - x_list[t - 1] + std::pow(dt,2) * a;
        v = (1 / dt) * (x - x_list[t]);
    }


  }
  auto t2 = std::chrono::high_resolution_clock::now();



  std::cout << "f() took "
      << std::chrono::duration_cast<std::chrono::milliseconds>(t2 - t1).count()
      << " milliseconds\n";

  // Write the trajectories to file
  ofstream fout;
  fout.open("trajectories.txt");
  if (fout) { // file opened successfully
    for (int i = 0; i < t_list.size(); i = i + 1) {
      fout << t_list[i] << ' ' << x_list[i] << ' ' << v_list[i] << endl;
    }
  } else { // file did not open successfully
    cout << "Could not open trajectory file for writing" << endl;
  }

  /* The file can be loaded and visualised in Python as follows:

  import numpy as np
  import matplotlib.pyplot as plt
  results = np.loadtxt('trajectories.txt')
  plt.figure(1)
  plt.clf()
  plt.xlabel('time (s)')
  plt.grid()
  plt.plot(results[:, 0], results[:, 1], label='x (m)')
  plt.plot(results[:, 0], results[:, 2], label='v (m/s)')
  plt.legend()
  plt.show()

  */
}
